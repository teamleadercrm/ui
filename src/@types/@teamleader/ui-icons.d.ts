declare module '@teamleader/ui-icons';
