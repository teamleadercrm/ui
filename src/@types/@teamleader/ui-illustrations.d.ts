declare module '@teamleader/ui-illustrations';
