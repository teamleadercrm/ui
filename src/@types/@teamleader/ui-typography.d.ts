declare module '@teamleader/ui-typography';
