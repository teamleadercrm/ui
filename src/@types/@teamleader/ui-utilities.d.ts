declare module '@teamleader/ui-utilities';
