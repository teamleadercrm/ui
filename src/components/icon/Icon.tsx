import React, { forwardRef, ReactNode } from 'react';
import Box from '../box';
import cx from 'classnames';
import theme from './theme.css';
import { BoxProps } from '../box/Box';
import { GenericComponent } from '../../@types/types';
import { COLORS, TINTS } from '../../constants';
import isComponentOfType from '../utils/is-component-of-type';

export interface IconProps extends Omit<BoxProps, 'children' | 'className'> {
  /** Element wrapped in Icon tags or sent as children prop */
  children?: ReactNode;
  /** Classname for the Icon component */
  className?: string;
  /** Color of the icon */
  color?: typeof COLORS[number];
  /** Tint of the icon */
  tint?: typeof TINTS[number];
  /** Opacity of the icon */
  opacity?: number;
}

const Icon: GenericComponent<IconProps> = forwardRef<HTMLElement, IconProps>(
  ({ children, className, color = 'teal', tint = 'normal', opacity = 0.84, ...others }, ref) => {
    const classNames = cx(theme[color], theme[tint], className);

    return (
      <Box
        className={classNames}
        alignItems="center"
        data-teamleader-ui="icon"
        display="flex"
        element="span"
        ref={ref}
        {...others}
      >
        {React.Children.map(children, (child) => {
          // Check if child is an actual React component
          // if so, pass the needed props. If not, just render it.
          if (!React.isValidElement(child) || !isComponentOfType(Icon, child)) {
            return child;
          }

          return React.cloneElement(child, {
            opacity,
          });
        })}
      </Box>
    );
  },
);

Icon.displayName = 'Icon';

export default Icon;
