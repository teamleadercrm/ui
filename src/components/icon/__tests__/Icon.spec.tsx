import React from 'react';
import { render } from '@testing-library/react';
import Icon from '../Icon';
import { IconWarningBadgedMediumOutline } from '@teamleader/ui-icons';

describe('Component - Icon', () => {
  it('renders', async () => {
    const { asFragment } = render(
      <Icon>
        <IconWarningBadgedMediumOutline />
      </Icon>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
