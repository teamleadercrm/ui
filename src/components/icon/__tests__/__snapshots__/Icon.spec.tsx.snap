// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - Icon renders 1`] = `
<DocumentFragment>
  <span
    class="box align-items-center display-flex teal normal"
    data-teamleader-ui="icon"
  >
    <svg
      fill="currentColor"
      height="24"
      preserveAspectRatio="xMidYMid meet"
      style="opacity: 0.84; vertical-align: middle;"
      viewBox="0 0 24 24"
      width="24"
    >
      <path
        clip-rule="evenodd"
        d="M12 22C6.477 22 2 17.523 2 12S6.477 2 12 2s10 4.477 10 10-4.477 10-10 10zm0-2a8 8 0 100-16 8 8 0 000 16zm0-1.975a1.5 1.5 0 11.001-3 1.5 1.5 0 01-.001 3zM12 6a1 1 0 011 1v6a1 1 0 11-2 0V7a1 1 0 011-1z"
        fill="currentColor"
        fill-rule="evenodd"
      />
    </svg>
  </span>
</DocumentFragment>
`;
